### ISSUE TITLE
Alignment Warning Caused by Leading Whitespace in Segment Text

### ISSUE THREAD

**Issue by LeMoussel** [kb-labb/easytranscriber#4]
During alignment, the following warning appears:
> Extracting emissions from audio files: 100%|█████████████████████████████████████████████| 54/54 [02:01<00:00,  2.25s/it]
>   0%|                                                                                                                                                                         | 0/54 [00:00<?, ?it/s]
> Segment indices start before the first token index. This may be due to leading whitespace in the original text. Consider stripping leading/trailing whitespace from the original text before creating SpeechSegment objects.

This warning likely originates from `easyaligner/alignment/pytorch.py`. It indicates that an alignment segment was created with a start index that precedes the first token index. In most cases, this happens because the original segment text contains leading spaces or newline characters.

The issue is generally non-fatal and does not stop the alignment process. However, it suggests that the input text spans are not fully normalized, which may lead to slightly inconsistent alignment boundaries.

A temporary workaround was implemented by adding a targeted deletion filter in the script to suppress the problematic segments.

A more robust solution would be to address the root cause directly in the library pipeline, by stripping leading and trailing whitespace from the text before creating `SpeechSegment` objects.

**Comment by Lauler** [comment]
Can you explain how you were using the package before this warning showed up? Is your script creating `SpeechSegment` objects, and adding `text` to those objects? 

My documentation is not specific enough about the usage of `SpeechSegment` objects. They have two main purposes:

1. Only transcribing/aligning a specific region of an audio file by specifying `start` and `end` for those regions. 
2. Inserting ground truth text (or ASR text) that corresponds to what is being spoken in that audio region, and force aligning that text with audio with custom/arbitrary segmentation of the text (i.e. can be paragraphs, sentences, or other granularity). This is the main purpose of the `easyaligner` package, and is better supported with the `pipeline()` function in that package. 

In general you don't need to use manually defined `SpeechSegment` objects in  `easytranscriber` for regular ASR. Whenever you use manually defined `SpeechSegment` objects, it is however important that you `.strip()` whitespace from the text. Otherwise the forced alignment may have inconsistent character boundaries, which may lead slightly wrong or offset timestamps. 

I'll be able to help you better if you share the settings/script you've used that produced this warning.

**Comment by joenaess** [comment]
I saw this issue and dug into the pipeline! While the root cause is likely the user's custom script, I just opened a PR which hardens ct2.py and hf.py to automatically construct and strip speech.text internally during the ASR phase to prevent this in the future. It also fixes a flat-list chunk indexing bug I found along the way. J

**Comment by Lauler** [comment]
Thanks @joenaess , there's indeed a bug in the indexing that will affect users which specify multiple `SpeechSegment` objects manually. Will look at your PR when I have time. 

Let me know if you are creating `SpeechSegment` objects @LeMoussel .

**Comment by LeMoussel** [comment]
No, I don't create `SpeechSegment` objects.

**MP3 Audio file:**
The MP3 file comes from this archive : [Emma (French Version)](https://librivox.org/emma-by-jane-austen-3/)

**Python code to reproduce:**
```python
from pathlib import Path

# https://github.com/kb-labb/easyaligner
from easyaligner.text import load_tokenizer

# https://github.com/kb-labb/easytranscriber
from easytranscriber.pipelines import pipeline
from easytranscriber.text.normalization import text_normalizer


data_dir = Path("./data/emma_1903_librivox")
audio_files = ["emma_02_austen_64kb.mp3"]

pipeline(
    vad_model="pyannote",
    # emissions_model (wav2vec2): Produces CTC logits for forced alignment only, Must match the audio input language
    # https://huggingface.co/jonatasgrosman/wav2vec2-large-xlsr-53-french
    emissions_model="jonatasgrosman/wav2vec2-large-xlsr-53-french",
    # transcription_model (Whisper): Converts audio → text, Must match the output language
    # https://huggingface.co/openai/whisper-large-v3-turbo
    transcription_model="openai/whisper-large-v3-turbo", 
    audio_paths=audio_files,
    audio_dir=str(data_dir),
    backend="ct2",
    language="fr",
    task="transcribe",
    tokenizer=load_tokenizer("french"),
    text_normalizer_fn=text_normalizer,
    cache_dir="models",
    output_vad_dir=str(data_dir / "output/vad"),
    output_transcriptions_dir=str(data_dir / "output/transcriptions"),
    output_emissions_dir=str(data_dir / "output/emissions"),
    output_alignments_dir=str(data_dir / "output/alignments"),
)
```

**Console Output:**
```text
Running VAD on audio files:   0%|                                                                                                          | 0/1 [00:00<?, ?it/s]/home/dev/Python/AI/Speech_Recognition/.venv/lib/python3.12/site-packages/easyaligner/vad/pyannote.py:434: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.detach().clone() or sourceTensor.detach().clone().requires_grad_(True), rather than torch.tensor(sourceTensor).
  "waveform": torch.tensor(audio).unsqueeze(0).to(torch.float32),
/home/dev/Python/AI/Speech_Recognition/.venv/lib/python3.12/site-packages/pyannote/audio/utils/reproducibility.py:74: ReproducibilityWarning: TensorFloat-32 (TF32) has been disabled as it might lead to reproducibility issues and lower accuracy.
It can be re-enabled by calling
   >>> import torch
   >>> torch.backends.cuda.matmul.allow_tf32 = True
   >>> torch.backends.cudnn.allow_tf32 = True
See https://github.com/pyannote/pyannote-audio/issues/1370 for more details.

  warnings.warn(
Running VAD on audio files: 100%|███████████████████████████████████| 1/1 [00:01<00:00,  1.49s/it]
Transcribing audio files: 100%|██████████████████████████████████████| 1/1 [00:03<00:00,  3.53s/it]
Loading weights: 100%|██████████████████████████████████████████| 424/424 [00:00<00:00, 5076.33it/s, Materializing param=wav2vec2.masked_spec_embed]
Extracting emissions from audio files: 100%|███████████████████████████████████████████████████████| 1/1 [00:01<00:00,  1.72s/it]
  0%|                                                                                                                                      | 0/1 [00:00<?, ?it/s]
Segment indices start before the first token index. This may be due to leading whitespace in the original text. Consider stripping leading/trailing whitespace from the original text before creating SpeechSegment objects.

100%|███████████████████████████████████████████████████| 1/1 [00:00<00:00,  1.78it/s]
```

**Comment by Lauler** [comment]
Thanks for posting a reproducible example. 

Turns out it's a false positive triggered by French guillemet quotation marks. The warning was only meant to warn about leading/trailing whitespace. But it's being triggered by the guillemet quotation marks due to how text normalization is being performed in my package. You can safely ignore the warning. 

I have patched the bug in `easyaligner` so the warning should not show up: https://github.com/kb-labb/easyaligner/commit/b3eb48377e4fa736b6ca9f2b739eaf165d16341b 

You can install the latest commit of `easyaligner` and the warning will disappear:

```
pip install git+https://github.com/kb-labb/easyaligner.git
```

Basically, `easyaligner` detects character indices for sentence/segment boundaries. For example below there are two sentences: 

```
Néanmoins, quand on servit le thé, il répéta exactement ce qu'il avait dit au déni. 
« Pauvre Mlle Taylor, que n'
```

The text is normalized at first to remove special characters 

```
Token: "déni"    → start_char=375, end_char=379
Token: "Pauvre"  → start_char=383, end_char=389
```

But the library also tries to keep track of the special characters that were deleted:

```
Token: "déni"    → start_char=375, end_char_extended=383, text_span_full='déni. « '
Token: "Pauvre"  → start_char=383, end_char_extended=390, text_span_full='Pauvre '
```

In this case it adds the ". «` to the word "déni". 

However, the sentence tokenizer thinks the second sentence starts at "«".

```
Sentence 1: (start=..., end=376)  "...il répéta exactement ce qu'il avait dit au déni."
Sentence 2: (start=381, end=...)  "« Pauvre Mlle Taylor, que n'..."
```

This created a mismatch between when the "segment index" (381) starts and when the first word of that segment starts (Pauvre starts at character 383). 

This mismatch does not matter in this case. It only causes problems when it happens at the very beginning of a chunk with leading whitespace.

**Comment by LeMoussel** [comment]
```bash
(.venv) dev@dev-B550M-DS3H:~/Python/AI/Speech_Recognition$ pip install -U git+https://github.com/kb-labb/easyaligner.git
Collecting git+https://github.com/kb-labb/easyaligner.git
  Cloning https://github.com/kb-labb/easyaligner.git to /tmp/pip-req-build-wg_r9o2a
  Running command git clone --filter=blob:none --quiet https://github.com/kb-labb/easyaligner.git /tmp/pip-req-build-wg_r9o2a
  Resolved https://github.com/kb-labb/easyaligner.git to commit b3eb48377e4fa736b6ca9f2b739eaf165d16341b
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
.....

(.venv) dev@dev-B550M-DS3H:~/Python/AI/Speech_Recognition$ pip freeze | grep easyaligner
easyaligner==0.2.0
```
The warning message is still present.
Found why 
`pip install -U git+https://github.com/kb-labb/easyaligner.git` didn't install the latest Git with the commit.

**Comment by LeMoussel** [comment]
It's OK with :
```bash
pip uninstall -y easyaligner
pip install easyaligner@git+https://github.com/kb-labb/easyaligner.git
```
